### PR TITLE
ECE and Fuel Rebalances

### PR DESCRIPTION
**Pull request by Steelux8**
These numbers are not final.

- Changed Biodiesel fuel value from 256k to 320k EU;
- Changed Benzene fuel value from 288k to 360k EU;
- Changed Cetane-Boosted Diesel fuel value from 720k to 1M EU;
- Changed High Octane Gasoline fuel value from 1.728M to 2.5M EU;
- Changed the recipe of Biodiesel to CBD to output 900L instead of 750L per 1000L of Biodiesel, to account for the fuel value change;
- Reduced the LOX consumption of the ECE from 320 to 80 L/s. A quad Vacuum Freezer setup, with 3 HV and 1  EV can feed one ECE;
- Small changes to the EU/t of the ECE due to it all being coded with ints in a specific format, where doubles caused several rounding errors. The boosted output amount is very similar to how it was before;
- Added a condition to the LCE to not burn HOG if it does not have Oxygen to boost it, since it generated power without consuming any fuel due to its consumption being smaller than 1 L/t, and therefore rounded down to 0.

After making [this spreadsheet regarding the HOG chain](https://docs.google.com/spreadsheets/d/13PetVb3BV11VojtpBslY4qtsNdAC61CXnWAr6y8wfVY/edit?usp=sharing), I realized that I would be doing it mainly for curiosity, given that this fuel, like many others, isn't worthwhile compared to Diesel, and also compared to the nuclear fission power available in EV/IV. After some conversation with other devs, I decided to make this PR to improve the fuel value of some fuels that are considered "main" paths for powergen, but that are also usually superceded by Diesel and how easy it is to generate a lot of power for very cheap and with a small setup, in comparison to these other fuels.

For Biodiesel and Benzene, they are held back by the need to set up a farm, usually an IC2 crop farm that needs manual breeding to multiply the crops needed to feed these setups in HV and above. CBD and HOG, on the other hand, require a lot more work to make and have a payoff that is too small compared to the processing chain built to automate them. I recognize that CBD might be in the best position out of these 4 fuels, before these changes, so I'm waiting to hear opinions from other people.

HOG also lacks good generators to burn it in. The ECE was made specifically for it, as it is the only fuel it accepts, but the LOX cost was obscene, a minimum of 9600 EU/t with HV Vacuum Freezers, with no overclocks. In this situation, you'd have to build 20 of them, and lose more than 1/4 of the power generated in the ECE. After the change, it will only need 80 LOX per second, manageable with a quad vac setup at 3360 EU/t, with one EV and three HV.

In my view, and the one from many others I've talked with before, a lot of these fuels are nice, but often disregarded due to the existence of Diesel and nukes. It'd be good to bring them closer to those two optimal powergen options.

### CODE REVIEW

**Comment by Steelux8** [comment]
I closed this because the previous PR I made for the Large Boiler Rebalance showed up here. I didn't make a new branch for this one, my mistake. This PR should be fine anyways, since the previous one is already merged. Please review it and share your view on the topic.

**Comment by bombcar** [comment]
Fuel | Old | New
-- | -- | --
Biodiesel | 256 | 320
Benzene | 288 | 360
CBD | 720 | 1000
HOG | 1728 | 2500

In addition, the CBD from biodiesel is increased from 750L CBD per 1000L of BD to 900L per 1000L.

**Comment by bombcar** [comment]
CBD went from 720 to 1000, I propose to not buff diesel further that the diesel-CBD recipe change to 

1000L Diesel + 20L tetranitromethane = 720L CBD

This means biodiesel and CBD get a buff without also buffing diesel any further

**Comment by querns** [comment]
**TL;DR: Maintain current TNM value add for (bio)diesel -> CBD:**

|Base Fuel      |Batch Size|TNM Amount|CBD Output Amount|
|---------------|----------|----------|-----------------|
|Diesel         |1000mB    |40mB      |1000mB           |
|Biodiesel (90%)|1000mB    |60mB      |900mB            |

The 1000mB diesel -> 720mB CBD would make its conversion ratio worse than biodiesel. Would it make more sense to adjust the tetranitromethane (TNM) amount instead?

For diesel, a 480,000 EU/B -> 720,000 EU/mB upgrade requires 20 mB of TNM. This means that each mB of TNM added to 1000mB of diesel adds 12,000 EU/mB. Extrapolating this to the new values, we would go from 480k EU/mB -> 1000k EU/mB, or a difference of 520k EU/mB. This works out to 43.333 mB of TNM. It could be rounded down to 40mB or rounded up to 44mB, depending on if you want clean numbers or more perfect accuracy.

Similarly, for biodiesel: The old recipe went from 256k EU/mB -> 648k EU/mB (assuming we keep the same 90% efficiency proposed.) This works out to a TNM value-add of 9.8k EU/mB. Translating this to the new proposition, we'd go from 320k EU/mB to 900k EU/mB, or a value-add of 580k EU/mB, necessitating 59.18 mB of TNM. 

Making these changes would make the numbers much cleaner, and maintain the value-add of TNM from the old system, while still providing a visible boost to biodiesel's conversion rate.

**Comment by bombcar** [comment]
I like the sound of that, and then why not just have both be 1000L -> 1000L? Makes it prettier all around (change the cost in TNM if you wish)

**Comment by Steelux8** [comment]
I think that's fine. I didn't want to change the CBD recipes too much, but I do want to give it a place, and it's hard to do it because it sits right next to diesel and doesn't have any multis dedicated to itself, unlike HOG.

**Comment by bombcar** [comment]
Agreed, most people build CBD because it's also rocket fuel (and they have diesel out their ears) - I'm not against it much as a buff just don't want diesel to get too good.

It's possible CBD goes away from being a rocket fuel at which point it would need to stand entirely on its own.

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java [hunk 1]
```diff
@@ -41,10 +41,10 @@ protected GT_Multiblock_Tooltip_Builder createTooltip() {
...
                 .addInfo("Controller block for the Extreme Combustion Engine")
                 .addInfo("Supply High Octane Gasoline and 8000L of Lubricant per hour to run")
-                .addInfo("Supply 320L/s of Liquid Oxygen to boost output (optional)")
-                .addInfo("Default: Produces 8192EU/t at 100% fuel efficiency")
-                .addInfo("Boosted: Produces 32768EU/t at 400% fuel efficiency")
-                .addInfo("You need to wait for it to reach 400% to output full power")
+                .addInfo("Supply 80L/s of Liquid Oxygen to boost output (optional)")
+                .addInfo("Default: Produces 10900EU/t at 100% fuel efficiency")
+                .addInfo("Boosted: Produces 32700EU/t at 150% fuel efficiency")
+                .addInfo("You need to wait for it to reach 300% to output full power")
```
Comment (bombcar) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: I still don't get this percentage - 300% fuel efficiency?

Comment (YannickMG) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: What I think I understand is that  supplying oxygen gives you 150% **fuel efficiency** which **somehow** allows the machine to go from 100% **machine efficiency** to 300% **machine efficiency**.

Comment (Steelux8) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: "Efficiency" was often used to show the speed of a machine. The tooltiip before these changes actually says that the boosted mode burns fuel at 400% efficiency, when it's 100%, the fuel value listen in NEI.

Comment (bombcar) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: Still confused but I think I get it - it gets 150% of the EU per cell when boosted, and at full power it does 3x *number of cells per minute* that it does when normally starting out?

Comment (Steelux8) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: No, it does 2x amount of cells as it would when unboosted, but it produces 3x the power: therefore, 150% efficiency.

Comment (bombcar) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:47]: Ah ok wow that is confusing lol

**Comment by 0lafe** [comment]
Imo HOG still needs a buff on the generator side of things, and mid IV is a weird spot for its setup, but these changes seem solid. 

Why buff benzene? I get that it's not great, but its not awful and the setup is trivial and infinite.

**Comment by bombcar** [comment]
I think to keep it comparable to biodiesel/light fuel.

**Comment by iouter** [comment]
now gg has also added two jet fuels, this should be rebalanced and allowed to burn in ECE.

**Comment by Steelux8** [comment]
> Why buff benzene? I get that it's not great, but its not awful and the setup is trivial and infinite.

How is the setup trivial? It needs more machines to fully automate it than light fuel does:

Crop Harvester -> Coke Oven -> Fluid Extractor -> Distillery (Benzene)
Pump -> Distillery -> Chemical Reactor (Light Fuel)

As a consequence, Light Fuel needs a smaller setup (no multis), runs faster than Benzene since both the coke ovens and fluid extractors run very slowly and need to be spammed, **and it skips the long process of crossbreeding bonsais to get wood automatically** by just pumping oil from places.

Oh, and Light Fuel has a 305k fuel value compared to Benzene's 288k, so you get rewarded with more EU per bucket for doing less work.

> Imo HOG still needs a buff on the generator side of things, and mid IV is a weird spot for its setup, but these changes seem solid.

Would you prefer seeing it in later tiers? I have thought of a change to the LuV Universal Chemical Engine to make its efficiency go down based on the amount of fluid it's consuming, instead of the energy produced. This would mean that the speed at which a fuel can produce power without losing efficiency depends on its fuel value, which makes HOG the king.

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java [hunk 2]
```diff
@@ -130,13 +130,11 @@ public IMetaTileEntity newMetaEntity(IGregTechTileEntity aTileEntity) {
 
     @Override
     protected int getNominalOutput() {
-        return 8192;
+        return 10900;
```
Comment (Glease) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:133]: 10900 is a really weird number. It makes you require a IV buffered dynamo instead of EV buffered dynamo. Not saying anyone would run this unboosted or this matters at all. It's just so weird.

Comment (Steelux8) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_ExtremeDieselEngine.java:133]: I agree that the numbers are weird, but it's an IV multi so people _should_ be running IV buffered dynamos anyways. I wanted to leave it as it was, but the code used all int variables and when I changed the variable types to change the fuel efficiency of the multi, I got wrong fuel consumption rates when testing because of rounding errors.

**Comment by riking** [comment]
> How is the setup trivial? It needs more machines to fully automate it than light fuel does:
> 
> Crop Harvester -> Coke Oven -> Fluid Extractor -> Distillery (Benzene)

Wood Tar doesn't need a fluid extractor? It's a direct output of the coke oven.

**Comment by querns** [comment]
> > How is the setup trivial? It needs more machines to fully automate it than light fuel does:
> > Crop Harvester -> Coke Oven -> Fluid Extractor -> Distillery (Benzene)
> 
> Wood Tar doesn't need a fluid extractor? It's a direct output of the coke oven.

You need a fluid extractor to process its charcoal output into additional wood tar.

**Review by Dream-Master — APPROVED** [review]

**Comment by Dream-Master** [comment]
lets test on zerta

**Comment by OrderedSet86** [comment]
These are all fair changes I think. I'm a fan of greater power generation choice variety.

Some comments on benzene meta:
1. Bonsais are not meta for benzene HV+. EnderIO farms and the new tree growth simulator are strictly better (no crop breeding, easier to scale). This means it's possible to go cropless for benzene.
2. The fluid extractor spam is very annoying/laggy at the moment due to it having two outputs and relatively low EU/t yield per machine. I personally prefer turning the charcoal into CO using oxygen in an LCR as it is easier to scale / upgrade, but this requires a robust oxygen input as the oxygen is 1:1 with CO.

Diagram for tree growth simulator + charcoal -> CO (using new numbers):
carbon_monoxide_power

Diagram for fluid extractors (using new numbers):
fluid_ex_benzene

**Comment by OrderedSet86** [comment]
Cetane buff seems a little unnecessary to me imho, it's already viable to use in runs for a long time post HV. In our current run we'll considering using cetane up to fusion and not even bothering with intermediate nukes, which means it's already quite good. But everything else makes sense to me.

**Comment by Steelux8** [comment]
> Cetane buff seems a little unnecessary to me imho, it's already viable to use in runs for a long time post HV. In our current run we're considering using pre-buff cetane up to fusion and not even bothering with intermediate nukes, which means it's already quite good. But everything else makes sense to me.

You could do this more easily if you just used diesel, though. If I am wrong about this, do show me that I am, but my view is that CBD is only 50% better in terms of fuel value than diesel, so if you just doubled your diesel setup, you'd end up with more power than making the tetranitromethane chain, and with less effort and expenses.

**Comment by OrderedSet86** [comment]
> Benzene starts at LV, and should be good enough to use at those early tiers.

Benzene is MV unless you're steam cracking heavy fuel for benzene, in which case you're doing something wrong (other oil products have higher EU yield).
image

> You could do this more easily if you just used diesel, though. If I am wrong about this, do show me that I am, but my view is that CBD is only 50% better in terms of fuel value than diesel, so if you just doubled your diesel setup, you'd end up with more power than making the tetranitromethane chain, and with less effort and expenses.

I think this is an extension of [the discussion we had earlier on Discord](https://discord.com/channels/181078474394566657/181078474394566657/931988616610205787). You're correct that scaling diesel is fewer machines. However, it requires significantly more oil for similar EU/t - it doesn't come free. Relevant charts:

diesel_mv
cetane_oil_dt

The second one is from querns and uses 50 oil/s to get similar yields as diesel, which is spending 125 oil/s. These are using the old cetane numbers. I think the tradeoff is fair between the two due to the additional costs of needing to set up additional oil drills and infrastructure to handle this (LDFPs, etc).

> To me, this is a problem with the benzene chain straight up not being good enough.

Correct, there is a reason fuel is meta. Benzene line QoL is not great. My goal in saying this was to imply other solutions to the benzene problem. For example, charcoal fluid extractor speed could be decreased - this is the main annoyance / machine spam source of the benzene line at the moment. Or a [cetane boosted benzene upgrade](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/8739#issuecomment-952282163) could be added so benzene has an HV upgrade (so you don't have to spam as hard to get to the mid-EV nitrobenzene transition).

**Comment by Steelux8** [comment]
> Benzene is MV unless you're steam cracking heavy fuel for benzene, in which case you're doing something wrong (other oil products have higher EU yield).

I forgot about this, somehow. MV distilleries are open basically at the start of MV, but that still means that the direct oil alternative to Benzene is Diesel, which is vastly superior. I'd give more value if it was a possibility at LV, but it is not the case.

> You're correct that scaling diesel is fewer machines. However, it requires significantly more oil for similar EU/t - it doesn't come free. The second one is from querns and uses 50 oil/s to get similar yields as diesel, which is spending 125 oil/s. These are using the old cetane numbers.

I agree that it's not free, but it still is cheaper than making CBD. To get this extra amount of oil to make more diesel, only one multi is required (the oil rig), and a few pipes to transport the fuel to the adjacent chunk. A discovered oil pool under bedrock is 8x8 chunks, so it's easy to just add another oil rig and a DT or two. CBD is unlocked later and needs more processing, but is it really better than just making diesel faster?

**Comment by OrderedSet86** [comment]
> I agree that it's not free, but it still is cheaper than making CBD. To get this extra amount of oil to make more diesel, only one multi is required (the oil rig), and a few pipes to transport the fuel to the adjacent chunk. A discovered oil pool under bedrock is 8x8 chunks, so it's easy to just add another oil rig and a DT or two. CBD is unlocked later and needs more processing, but is it really better than just making diesel faster?

One added wrinkle to this is that CBD is still T1 T2 rocket fuel (unless the LCR rocket fuel removal merges), so you're going to end up making it anyway. So if you already have the tetranitromethane, might as well use LCRs...

**Comment by bombcar** [comment]
I'm 💯 in the "I made CBD too get moon/mars" but never actually finalized automating it, and diesel still runs most my base and further power is likely to be nuclear unless I meme CBD for Chemplant flex.
